File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/SeekBarChangeEventObservable.java
package com.jakewharton.rxbinding2.widget;

import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SeekBarChangeEventObservable extends InitialValueObservable<SeekBarChangeEvent> {
  private final SeekBar view;

  SeekBarChangeEventObservable(SeekBar view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super SeekBarChangeEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    view.setOnSeekBarChangeListener(listener);
    observer.onSubscribe(listener);
  }

  @Override protected SeekBarChangeEvent getInitialValue() {
    return new SeekBarProgressChangeEvent(view, view.getProgress(), false);
  }

  static final class Listener extends MainThreadDisposable implements OnSeekBarChangeListener {
    private final SeekBar view;
    private final Observer<? super SeekBarChangeEvent> observer;

    Listener(SeekBar view, Observer<? super SeekBarChangeEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
      if (!isDisposed()) {
        observer.onNext(new SeekBarProgressChangeEvent(seekBar, progress, fromUser));
      }
    }

    @Override public void onStartTrackingTouch(SeekBar seekBar) {
      if (!isDisposed()) {
        observer.onNext(new SeekBarStartChangeEvent(seekBar));
      }
    }

    @Override public void onStopTrackingTouch(SeekBar seekBar) {
      if (!isDisposed()) {
        observer.onNext(new SeekBarStopChangeEvent(seekBar));
      }
    }

    @Override protected void onDispose() {
      view.setOnSeekBarChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/widget/SeekBarChangeEventObservable.kt
@file:JvmName("RxSeekBar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.widget

import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of progress change events for `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun SeekBar.changeEvents(): InitialValueObservable<SeekBarChangeEvent> {
  return SeekBarChangeEventObservable(this)
}

private class SeekBarChangeEventObservable(
  private val view: SeekBar
) : InitialValueObservable<SeekBarChangeEvent>() {

  override fun subscribeListener(observer: Observer<in SeekBarChangeEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.setOnSeekBarChangeListener(listener)
    observer.onSubscribe(listener)
  }

  override val initialValue get() = SeekBarProgressChangeEvent(view, view.progress, false)

  private class Listener(
    private val view: SeekBar,
    private val observer: Observer<in SeekBarChangeEvent>
  ) : MainThreadDisposable(), OnSeekBarChangeListener {

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
      if (!isDisposed) {
        observer.onNext(SeekBarProgressChangeEvent(seekBar, progress, fromUser))
      }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
      if (!isDisposed) {
        observer.onNext(SeekBarStartChangeEvent(seekBar))
      }
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
      if (!isDisposed) {
        observer.onNext(SeekBarStopChangeEvent(seekBar))
      }
    }

    override fun onDispose() {
      view.setOnSeekBarChangeListener(null)
    }
  }
}
